backend/src/nodes/utils/ncnn_optimizer.py

```diff
@@ -0,0 +1,1209 @@
+                        j += 1
+
+                    if j == len(self.model.layers):
+                        continue
+
+                    convolution = self.model.layers[j]
+                    innerproduct2 = NcnnLayer(
+                        "InnerProduct",
+                        convolution.name,
+                        convolution.num_inputs,
+                        convolution.num_outputs,
+                        convolution.inputs,
+                        convolution.outputs,
+                    )
+                    innerproduct2.add_param(
+                        0, checked_cast(int, convolution.params[0].value)
+                    )
+                    innerproduct2.add_param(
+                        1, checked_cast(int, convolution.params[5].value)
+                    )
+                    innerproduct2.add_param(
+                        2, checked_cast(int, convolution.params[6].value)
+                    )
+                    innerproduct2.add_param(
+                        8, checked_cast(int, convolution.params[8].value)
+                    )
+                    innerproduct2.add_param(
+                        9, checked_cast(int, convolution.params[9].value)
+                    )
+                    innerproduct2.add_param(
+                        10,
+                        checked_cast(list, convolution.params[10].value),
+                    )
+                    innerproduct2.add_weight(
+                        "weight",
+                        convolution.weight_data["weight"].weight,
+                        convolution.weight_data["weight"].quantize_tag,
+                    )
+                    innerproduct2.add_weight(
+                        "bias", convolution.weight_data["bias"].weight
+                    )
+
+                    self.model.layers[j] = innerproduct2
+
+                    replaced = True
+
+            if not replaced:
+                break
+
+    def optimize(self):
+        self.__fuse_batchnorm_scale()
+        self.__fuse_x_batchnorm()
+        self.__fuse_x_mul()
+        self.__fuse_x_add()
+        self.__fuse_innerproduct_dropout()
+
+        self.__replace_reduction_with_global_pooling()
+        self.__replace_prelu_with_leaky_relu()
+
+        self.__fuse_x_activation()
+        self.__fuse_memorydata_binaryop()
+        self.__fuse_binaryop_eltwise()
+
+        self.__eliminate_dropout()
+        self.__eliminate_pooling1x1()
+        self.__eliminate_noop()
+        self.__eliminate_split()
+        self.__eliminate_flatten_after_global_pooling()
+        self.__eliminate_reshape_after_global_pooling()
+        self.__eliminate_reshape_before_binaryop()
+
+        self.__replace_convolution_with_innerproduct_after_global_pooling()
+        self.__replace_convolution_with_innerproduct_after_innerproduct()
+
+        self.__eliminate_flatten_after_innerproduct()
+        self.__eliminate_orphaned_memorydata()
+
+        return self.model
```

Comment:
Not a fan of this API. This makes it seem like you are returning a new optimized model, but that's not what's going on.

Reply:
Same question as before: should it copy the model before modifying it?

Reply:
(Let's talk about cache mutation and API here.)

> should it copy the model before modifying it?

Yes and no. In the case of Save Model, a copy is necessary for correctness, but in ONNX->NCNN the copy is unnecessary since we just created that model.

Maybe we could split the API into `def optimize(self) -> None` and `def optimized(self) -> NcnnModel`. This should make it clearer which method mutates the underlying model and which one doesn't (I hope). Maybe we could even call it `optimized_copy`?

Reply:
I'm going to hold off on this until we definitely settle on when it will be used. I don't particularly like the idea, but if we do decide to put it in Load, then I'd think mutating the model would be fine.

Reply:
So it should be fine if it always mutates the cached model now, yes? Since the optimized model is now always passed out of Load and Convert?

Reply:
Not really. The API of this function is still confusing.

Reply:
When I asked if it should always copy before though, you said it was unnecessary because the model had just been created, which is the same for Load Model.

Reply:
I changed the Optimizer so it now returns None.